### DISCUSSION TITLE
`DatePicker` within a form

### DISCUSSION THREAD

**Discussion by mrodrigues95** [adobe/react-spectrum#3154] [Q&A] [answered]
I've searched the storybook and couldn't find any examples of this.

Are there any examples on how to get the `DatePicker` wired up within a form? I've used a similar approach to https://react-spectrum.adobe.com/react-aria/useDatePicker.html with `react-hook-form` and managed to get it somewhat working although it feels very hacky so I'm curious if there is an easy way to get this hooked up.

I'm currently using `react-hook-form` along with `zod` as the resolver for form validation.

**Comment by ktabors** [comment]
Do you have a codesandbox of what you've tried? What part of it or in what way does it feel hacky?

**Comment by snowystinger** [comment, marked as answer]
We have a few discussions and issues where we've talked about usage with react-hook-form, are any of these able to help resolve your issues?
https://github.com/adobe/react-spectrum/issues/2603
https://github.com/adobe/react-spectrum/issues/1860#issuecomment-840125913

**Reply by mrodrigues95** [reply]
I managed to get something put together that's working decently well. The biggest issue I had was performing form validation and dealing with error states.

`react-hook-form` expects you to [forward](https://react-hook-form.com/api/usecontroller) a `ref` when controlled so that when an error occurs, the field will be focused. This is typically an `input` of some sort so I'm not sure how to get this hooked up with `DatePicker` given how that component is being rendered.

Another thing that seems odd is `useController` -> `onChange` spreads the `value` which causes `CalendarDate` to lost its inheritance meaning I have to construct a new instance every time the `value` changes to avoid errors like `toDate() is not a function`.

I don't have a direct sandbox that illustrates this implementation but you can view the above implementation at:
- https://github.com/mrodrigues95/spout/blob/main/libs/toolkit/src/lib/form/form-datepicker.tsx
- https://github.com/mrodrigues95/spout/blob/main/libs/toolkit/src/lib/form/form.stories.tsx

**Reply by snowystinger** [reply]
Easiest thing for us to help you is to make a codesandbox. I think this line of code from our own implementation can help with your focus through forwarded ref https://github.com/adobe/react-spectrum/blob/main/packages/%40react-spectrum/datepicker/src/DatePicker.tsx#L57 you should be able to do something similar.

**Reply by mrodrigues95** [reply]
`useFocusManagerRef` implementation you linked seems to work well with my use case. I've managed to get everything working in a decent manner now. Thank you for the help.

We can mark this discussion as answered.
